// the-component-util/shim/index.js
// Code generated by coz. DO NOT EDIT.

/**
 * Utility for the-components
 * @module the-component-util
 */
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "asHOC", {
  enumerable: true,
  get: function get() {
    return _asHOC.default;
  }
});
Object.defineProperty(exports, "asStyleData", {
  enumerable: true,
  get: function get() {
    return _asStyleData.default;
  }
});
Object.defineProperty(exports, "changedProps", {
  enumerable: true,
  get: function get() {
    return _changedProps.default;
  }
});
Object.defineProperty(exports, "colorAlpha", {
  enumerable: true,
  get: function get() {
    return _colorAlpha.default;
  }
});
Object.defineProperty(exports, "eventHandlersFor", {
  enumerable: true,
  get: function get() {
    return _eventHandlersFor.default;
  }
});
Object.defineProperty(exports, "htmlAttributesFor", {
  enumerable: true,
  get: function get() {
    return _htmlAttributesFor.default;
  }
});
Object.defineProperty(exports, "isKeyCode", {
  enumerable: true,
  get: function get() {
    return _isKeyCode.default;
  }
});
Object.defineProperty(exports, "isServerSide", {
  enumerable: true,
  get: function get() {
    return _isServerSide.default;
  }
});
Object.defineProperty(exports, "isVideoSrc", {
  enumerable: true,
  get: function get() {
    return _isVideoSrc.default;
  }
});
Object.defineProperty(exports, "newId", {
  enumerable: true,
  get: function get() {
    return _newId.default;
  }
});
Object.defineProperty(exports, "readFileAsDataURL", {
  enumerable: true,
  get: function get() {
    return _readFileAsDataURL.default;
  }
});
Object.defineProperty(exports, "toggleBodyClass", {
  enumerable: true,
  get: function get() {
    return _toggleBodyClass.default;
  }
});
Object.defineProperty(exports, "wrapStack", {
  enumerable: true,
  get: function get() {
    return _wrapStack.default;
  }
});

var _asHOC = _interopRequireDefault(require("./asHOC"));

var _asStyleData = _interopRequireDefault(require("./asStyleData"));

var _changedProps = _interopRequireDefault(require("./changedProps"));

var _colorAlpha = _interopRequireDefault(require("./colorAlpha"));

var _eventHandlersFor = _interopRequireDefault(require("./eventHandlersFor"));

var _htmlAttributesFor = _interopRequireDefault(require("./htmlAttributesFor"));

var _isKeyCode = _interopRequireDefault(require("./isKeyCode"));

var _isServerSide = _interopRequireDefault(require("./isServerSide"));

var _isVideoSrc = _interopRequireDefault(require("./isVideoSrc"));

var _newId = _interopRequireDefault(require("./newId"));

var _readFileAsDataURL = _interopRequireDefault(require("./readFileAsDataURL"));

var _toggleBodyClass = _interopRequireDefault(require("./toggleBodyClass"));

var _wrapStack = _interopRequireDefault(require("./wrapStack"));
