// the-check/shim/index.js
// Code generated by coz. DO NOT EDIT.

/**
 * Check utility for the-framework
 * @module the-check
 */
'use strict';

var canTouch_ = require('./canTouch');

var hasBin_ = require('./hasBin');

var isBrowser_ = require('./isBrowser');

var isChrome_ = require('./isChrome');

var isFirefox_ = require('./isFirefox');

var isMacOS_ = require('./isMacOS');

var isProduction_ = require('./isProduction');

var isiOS_ = require('./isiOS');

var unlessProduction_ = require('./unlessProduction'); // `module.exports` overrides these `exports.*`, but still needs them for lebab (https://github.com/lebab/lebab)


exports.canTouch = canTouch_;
exports.hasBin = hasBin_;
exports.isBrowser = isBrowser_;
exports.isChrome = isChrome_;
exports.isFirefox = isFirefox_;
exports.isMacOS = isMacOS_;
exports.isProduction = isProduction_;
exports.isiOS = isiOS_;
exports.unlessProduction = unlessProduction_;
module.exports = {
  canTouch: canTouch_,
  hasBin: hasBin_,
  isBrowser: isBrowser_,
  isChrome: isChrome_,
  isFirefox: isFirefox_,
  isMacOS: isMacOS_,
  isProduction: isProduction_,
  isiOS: isiOS_,
  unlessProduction: unlessProduction_
};
