// the-spin/shim/esm/TheSpinStyle.js
'use strict';

import _extends from "@babel/runtime/helpers/extends";
import c from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { asStyleData } from 'the-component-util';
import { TheStyle } from 'the-style';
/** Style for TheSpin */

var TheSpinStyle = function TheSpinStyle(_ref) {
  var className = _ref.className,
      id = _ref.id,
      options = _ref.options;
  return React.createElement(TheStyle, _extends({
    id: id
  }, {
    className: c('the-spin-style', className),
    styles: TheSpinStyle.data(options)
  }));
};

TheSpinStyle.displayName = 'TheSpinStyle';
TheSpinStyle.propTypes = {
  /** Style options */
  options: PropTypes.object
};
TheSpinStyle.defaultProps = {
  options: {}
};

TheSpinStyle.data = function (options) {
  var ThemeValues = TheStyle.ThemeValues;
  var _options$dominantColo = options.dominantColor,
      dominantColor = _options$dominantColo === void 0 ? ThemeValues.dominantColor : _options$dominantColo,
      _options$lightTextCol = options.lightTextColor,
      lightTextColor = _options$lightTextCol === void 0 ? ThemeValues.lightTextColor : _options$lightTextCol,
      _options$overlayBackg = options.overlayBackgroundColor,
      overlayBackgroundColor = _options$overlayBackg === void 0 ? ThemeValues.overlayBackgroundColor : _options$overlayBackg;
  return asStyleData({
    '.the-spin': {
      display: 'none',
      textAlign: 'center'
    },
    '.the-spin.the-spinner-cover': {
      alignItems: 'center',
      background: overlayBackgroundColor,
      bottom: 0,
      color: lightTextColor,
      justifyContent: 'center',
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
      zIndex: 4
    },
    '.the-spin.the-spinner-enabled': {
      alignItems: 'center',
      display: 'inline-flex',
      height: '100%',
      justifyContent: 'center',
      width: '100%'
    },
    '.the-spinner-icon': {
      display: 'inline-block',
      margin: '0 4px',
      opacity: 1,
      transition: 'opacity 100ms'
    }
  });
};

export default TheSpinStyle;
