// the-spin/shim/esm/TheSpin.js
'use strict';

import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import c from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { eventHandlersFor, htmlAttributesFor } from 'the-component-util';
import { TheIcon } from 'the-icon';
import TheSpinStyle from './TheSpinStyle';
var SpinnerSizes = ['medium', 'xx-small', 'x-small', 'small', 'large', 'x-large', 'xx-large', 'smaller', 'larger'];
var DEFAULT_THEME = 'C';
var DEFAULT_SIZE = 'large';
/**
 * Spin of the-components
 */

var TheSpin =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(TheSpin, _React$PureComponent);

  function TheSpin() {
    _classCallCheck(this, TheSpin);

    return _possibleConstructorReturn(this, _getPrototypeOf(TheSpin).apply(this, arguments));
  }

  _createClass(TheSpin, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var children = props.children,
          className = props.className,
          cover = props.cover,
          enabled = props.enabled,
          size = props.size,
          theme = props.theme;
      return React.createElement("div", _extends({}, htmlAttributesFor(props, {
        except: ['className']
      }), eventHandlersFor(props, {
        except: []
      }), {
        className: c('the-spin', className, {
          'the-spinner-cover': !!cover,
          'the-spinner-enabled': !!enabled
        })
      }), React.createElement(TheIcon.Spin, {
        className: "the-spin-icon",
        style: {
          fontSize: size
        },
        theme: theme
      }), children);
    }
  }]);

  return TheSpin;
}(React.PureComponent);

TheSpin.Style = TheSpinStyle;
TheSpin.Sizes = SpinnerSizes;
TheSpin.propTypes = {
  /** Shows spin */

  /** show as cover */
  cover: PropTypes.bool,
  enabled: PropTypes.bool,

  /** Size of spinner icon */
  size: PropTypes.oneOf(SpinnerSizes),

  /** Theme of the spinner */
  theme: PropTypes.oneOf(Object.keys(TheIcon.SpinIconThemes))
};
TheSpin.defaultProps = {
  cover: false,
  enabled: false,
  size: DEFAULT_SIZE,
  theme: DEFAULT_THEME
};
TheSpin.displayName = 'TheSpin';
export default TheSpin;
