// the-spin/shim/esm/withSpin.js
/**
 * HOC for spin
 * @function withSpin
 */
'use strict';

import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import { clone } from 'asobj';
import PropTypes from 'prop-types';
import React from 'react';
import TheSpin from './TheSpin';
/** @lends withSpin */

function withSpin(Component) {
  var WithSpin =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(WithSpin, _React$Component);

    function WithSpin() {
      _classCallCheck(this, WithSpin);

      return _possibleConstructorReturn(this, _getPrototypeOf(WithSpin).apply(this, arguments));
    }

    _createClass(WithSpin, [{
      key: "render",
      value: function render() {
        var props = this.props;
        var children = props.children,
            _props$spinner = props.spinner,
            spinner = _props$spinner === void 0 ? React.createElement(TheSpin, {
          cover: true,
          enabled: spinning,
          size: spinnerSize,
          theme: spinnerTheme
        }) : _props$spinner,
            spinnerSize = props.spinnerSize,
            spinnerTheme = props.spinnerTheme,
            spinning = props.spinning;
        var componentProps = clone(props, {
          without: Object.keys(WithSpin.propTypes)
        });
        return React.createElement(Component, _extends({}, componentProps, {
          spinner: spinner
        }), children);
      }
    }]);

    return WithSpin;
  }(React.Component);

  WithSpin.propTypes = {
    /** Show loading spinner */

    /** Spinner instance */
    spinner: PropTypes.node,

    /** Size of spinner */
    spinnerSize: PropTypes.string,

    /** Theme of the spinner */
    spinnerTheme: PropTypes.string,
    spinning: PropTypes.bool
  };
  return WithSpin;
}

export default withSpin;
