// the-button/shim/TheButtonGroup.js
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _theComponentUtil = require("the-component-util");

/**
 * Group of buttons
 */
var TheButtonGroup =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(TheButtonGroup, _React$Component);

  function TheButtonGroup() {
    (0, _classCallCheck2.default)(this, TheButtonGroup);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TheButtonGroup).apply(this, arguments));
  }

  (0, _createClass2.default)(TheButtonGroup, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var align = props.align,
          children = props.children,
          className = props.className,
          collapsed = props.collapsed,
          nowrap = props.nowrap;
      return _react.default.createElement("div", (0, _extends2.default)({}, (0, _theComponentUtil.htmlAttributesFor)(props, {
        except: ['className']
      }), (0, _theComponentUtil.eventHandlersFor)(props, {
        except: []
      }), {
        className: (0, _classnames.default)('the-button-group', className, {
          'the-button-group-center': align === 'center',
          'the-button-group-collapsed': collapsed,
          'the-button-group-left': align === 'left',
          'the-button-group-nowrap': nowrap,
          'the-button-group-right': align === 'right'
        })
      }), children);
    }
  }]);
  return TheButtonGroup;
}(_react.default.Component);

TheButtonGroup.propTypes = {
  /** collapsed style */

  /** Direction to align */
  align: _propTypes.default.oneOf(['left', 'right', 'center']),
  collapsed: _propTypes.default.bool
};
TheButtonGroup.defaultProps = {
  align: 'center',
  collapsed: false
};
TheButtonGroup.displayName = 'TheButtonGroup';
var _default = TheButtonGroup;
exports.default = _default;
