// liowebrtc/dist/liowebrtc.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _wildemitter = require('wildemitter');

var _wildemitter2 = _interopRequireDefault(_wildemitter);

var _attachmediastream = require('attachmediastream');

var _attachmediastream2 = _interopRequireDefault(_attachmediastream);

var _mockconsole = require('mockconsole');

var _mockconsole2 = _interopRequireDefault(_mockconsole);

var _webrtc = require('./webrtc');

var _webrtc2 = _interopRequireDefault(_webrtc);

var _webrtcsupport = require('./webrtcsupport');

var _webrtcsupport2 = _interopRequireDefault(_webrtcsupport);

var _socketioconnection = require('./socketioconnection');

var _socketioconnection2 = _interopRequireDefault(_socketioconnection);

var _PeerOptimizer = require('./PeerOptimizer');

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LioWebRTC = function (_WildEmitter) {
  _inherits(LioWebRTC, _WildEmitter);

  function LioWebRTC(opts) {
    _classCallCheck(this, LioWebRTC);

    var _this = _possibleConstructorReturn(this, (LioWebRTC.__proto__ || Object.getPrototypeOf(LioWebRTC)).call(this));

    var self = _this;
    var options = opts || {};
    _this.config = _constants.defaultConfig;
    var config = _this.config;

    _this.peerDataCache = {};
    _this.unconnectivePeers = {};
    _this.id = '';
    _this.roomCount = 0;
    _this.roomName = '';

    var connection = void 0;
    // Set up logger
    _this.logger = function () {
      if (opts.debug) {
        return opts.logger || console;
      }
      return opts.logger || _mockconsole2.default;
    }();

    // Set our config from options
    Object.keys(options).forEach(function (o) {
      _this.config[o] = options[o];
    });

    if (options.dataOnly) {
      _this.config.media.video = false;
      _this.config.media.audio = false;
      _this.config.receiveMedia.offerToReceiveAudio = false;
      _this.config.receiveMedia.offerToReceiveVideo = false;
    }

    if (!_this.config.media.video && _this.config.media.audio) {
      _this.config.localVideo.audio = true;
    }
    _this.capabilities = _webrtcsupport2.default;
    if (_this.config.connection === null) {
      connection = _this.connection = new _socketioconnection2.default(_this.config);
    } else {
      connection = _this.connection = _this.config.connection;
    }

    connection.on('connect', function () {
      self.emit('connectionReady', connection.getSessionid());
      self.sessionReady = true;
      self.testReadiness();
    });

    connection.on('message', function (message) {
      var peers = self.webrtc.getPeers(message.from, message.roomType);
      var totalPeers = self.webrtc.getPeers().length;
      var peer = void 0;

      if (message.type === 'offer') {
        if (peers.length) {
          peers.forEach(function (p) {
            if (p.sid === message.sid) peer = p;
          });
          // if (!peer) peer = peers[0]; // fallback for old protocol versions
        }
        if (_this.config.dataOnly && _this.config.network.maxPeers > 0 && totalPeers >= _this.config.network.maxPeers) {
          return;
        }
        if (!peer) {
          peer = self.webrtc.createPeer({
            id: message.from,
            sid: message.sid,
            type: message.roomType,
            enableDataChannels: self.config.enableDataChannels,
            sharemyscreen: message.roomType === 'screen' && !message.broadcaster,
            broadcaster: message.roomType === 'screen' && !message.broadcaster ? self.connection.getSessionid() : null
          });
          _this.sendPing(peer, peer.id, true);
        } else {
          return;
        }
        peer.handleMessage(message);
      } else if (peers.length) {
        peers.forEach(function (p) {
          p.handleMessage(message);
        });
      }
    });

    connection.on('remove', function (room) {
      if (room.id !== self.connection.getSessionid()) {
        self.webrtc.removePeers(room.id, room.type);
      }
    });

    opts.logger = _this.logger;
    opts.debug = false;
    _this.webrtc = new _webrtc2.default(opts);
    _constants.inheritedMethods.forEach(function (method) {
      self[method] = self.webrtc[method].bind(self.webrtc);
    });

    // proxy events from WebRTC
    _this.webrtc.on('*', function () {
      // eslint-disable-line
      self.emit.apply(self, arguments); // eslint-disable-line
    });

    // log all events in debug mode
    if (config.debug) {
      _this.on('*', _this.logger.log.bind(_this.logger, 'LioWebRTC event:'));
    }

    // check for readiness
    _this.webrtc.on('localStream', function () {
      self.testReadiness();
    });

    _this.webrtc.on('message', function (payload) {
      self.connection.emit('message', payload);
    });

    _this.webrtc.on('peerStreamAdded', _this.handlePeerStreamAdded.bind(_this));
    _this.webrtc.on('peerStreamRemoved', _this.handlePeerStreamRemoved.bind(_this));

    // echo cancellation attempts
    if (_this.config.adjustPeerVolume) {
      _this.webrtc.on('speaking', _this.setVolumeForAll.bind(_this, _this.config.peerVolumeWhenSpeaking));
      _this.webrtc.on('stoppedSpeaking', _this.setVolumeForAll.bind(_this, 1));
    }

    connection.on('stunservers', function (args) {
      // resets/overrides the config
      self.webrtc.config.peerConnectionConfig.iceServers = args;
      self.emit('stunservers', args);
    });
    connection.on('turnservers', function (args) {
      // appends to the config
      self.webrtc.config.peerConnectionConfig.iceServers = self.webrtc.config.peerConnectionConfig.iceServers.concat(args);
      self.emit('turnservers', args);
    });
    /*
    this.webrtc.on('iceFailed', (peer) => {
      // local ice failure
    });
    this.webrtc.on('connectivityError', (peer) => {
      // remote ice failure
    });
    */

    // sending mute/unmute to all peers
    _this.webrtc.on('audioOn', function () {
      self.webrtc.sendToAll('unmute', { name: 'audio' });
    });
    _this.webrtc.on('audioOff', function () {
      self.webrtc.sendToAll('mute', { name: 'audio' });
    });
    _this.webrtc.on('videoOn', function () {
      self.webrtc.sendToAll('unmute', { name: 'video' });
    });
    _this.webrtc.on('videoOff', function () {
      self.webrtc.sendToAll('mute', { name: 'video' });
    });

    self.on('removedPeer', function (peer) {
      if (peer.id) {
        (0, _PeerOptimizer.removeConnection)(_this.id, peer.id);
      }
    });

    self.on('channelClose', function (channel) {
      if (channel.label === 'liowebrtc' && _this.config.dataOnly && _this.config.network.maxPeers > 0 && (0, _PeerOptimizer.getNeighbors)(_this.id).length < _this.config.network.minPeers) {
        _this.connectToRandomPeer();
      }
    });

    _this.webrtc.on('channelMessage', function (peer, label, data) {
      if (data.payload._id && _this.peerDataCache[data.payload._id]) {
        return;
      }
      switch (data.type) {
        case '_volume':
          self.emit('remoteVolumeChange', data.payload, peer);
          break;
        case '_propagate':
          if (_this.seenPeerEvent(data.payload._id)) {
            return;
          }
          // Re-propagate message
          _this.propagateMessage(data.payload);
          _this.cachePeerEvent(data.payload._id, data.payload.senderId);
          // Emit the propagated data as if it were received directly
          self.emit('receivedPeerData', data.payload.type, data.payload.payload, {
            id: data.payload.senderId,
            nick: data.payload.senderNick,
            isForwarded: true
          });
          break;
        case '_ping':
          _this.sendPong(peer, data.payload);
          break;
        case '_pong':
          (0, _PeerOptimizer.addConnection)(_this.id, peer.id, Date.now() - data.payload[0] + data.payload[1]);
          break;
        case '_connections':
          data.payload.forEach(function (connection) {
            return (0, _PeerOptimizer.addConnection)(peer.id, connection.id, connection.weight);
          });
          break;
        default:
          if (_this.seenPeerEvent(data._id)) {
            return;
          }
          _this.cachePeerEvent(data._id, peer.id);
          self.emit('receivedPeerData', data.type, data.payload, peer);
          if (_this.config.network.maxPeers > 0 && data.shout) {
            data.senderId = peer.id;
            var fwdData = Object.assign({}, { senderId: peer.id, senderNick: peer.nick }, data);
            _this.propagateMessage(fwdData);
          }
          break;
      }
    });

    if (_this.config.autoRequestMedia) _this.startLocalVideo();
    return _this;
  }

  _createClass(LioWebRTC, [{
    key: 'cachePeerEvent',
    value: function cachePeerEvent(eventId, peerId) {
      if (!this.peerDataCache[eventId]) {
        this.peerDataCache[eventId] = {
          recipients: _defineProperty({}, peerId, true),
          timestamp: Date.now()
        };
        return;
      }
      if (!this.peerDataCache[eventId].recipients[peerId]) {
        this.peerDataCache[eventId].recipients[peerId] = true;
      }
      if (Object.keys(this.peerDataCache).length > 1024) {
        // Sort by timestamp
        var sortedCache = Object.entries(this.peerDataCache).sort(function (a, b) {
          return a[1].timestamp - b[1].timestamp;
        });
        // Delete oldest item
        delete this.peerDataCache[sortedCache[0][0]];
      }
    }
  }, {
    key: 'seenPeerEvent',
    value: function seenPeerEvent(eventId) {
      if (this.peerDataCache[eventId]) {
        return true;
      }
      return false;
    }
  }, {
    key: 'sendPong',
    value: function sendPong(peer, start) {
      var channel = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _constants.defaultChannel;

      var now = Date.now();
      peer.sendDirectly('_pong', [now, now - start], channel);
    }
  }, {
    key: 'sendPing',
    value: function sendPing(peer, peerId) {
      var firstPing = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var channel = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : _constants.defaultChannel;

      if (firstPing) peer.start();
      setTimeout(this.ping.bind(this, peer, peerId, firstPing, channel), 1000);
    }
  }, {
    key: 'ping',
    value: function ping(peer, peerId, firstPing, channel) {
      var tries = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 0;

      if (peer.sendDirectly('_ping', Date.now(), channel)) {
        // this.logger.log('sent ping to', peer.id);
        if (firstPing) this.emit('createdPeer', peer);
      } else {
        // The channel is closed
        if (tries === 2) {
          this.unconnectivePeers[peerId] = true;
          peer.end(false);
          return;
        }
        setTimeout(this.ping.bind(this, peer, peerId, firstPing, channel, tries + 1), 1000);
      }
    }
  }, {
    key: 'connectToRandomPeer',
    value: function connectToRandomPeer() {
      var _this2 = this;

      console.log('connecting rando');
      this.getClients(function (err, clients) {
        var ids = Object.keys(clients).filter(function (c) {
          return !(_this2.unconnectivePeers[c] === true || c === _this2.id || (0, _PeerOptimizer.isNeighbor)(_this2.id, c));
        });
        if (ids.length) {
          var randId = ids[Math.floor(Math.random() * ids.length)];
          _this2.connectToPeer(randId, clients[randId]);
        }
      });
    }
  }, {
    key: 'sendConnections',
    value: function sendConnections(peer) {
      var _this3 = this;

      var channel = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _constants.defaultChannel;

      if (peer.sendDirectly('_connections', this.getPeers().map(function (p) {
        var edge = _PeerOptimizer.Graph.findEdge(_this3.id, p.id);
        return { id: p.id, weight: edge.getWeight() };
      }), channel)) {
        // connections sent
      } else {
        peer.end();
      }
    }
  }, {
    key: 'propagateMessage',
    value: function propagateMessage(data) {
      var _this4 = this;

      var channel = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _constants.defaultChannel;

      this.getPeers().forEach(function (peer) {
        if (!_this4.peerDataCache[data._id]) {
          _this4.cachePeerEvent(data._id, data.senderId);
        }
        if (!_this4.peerDataCache[data._id].recipients[peer.id]) {
          peer.sendDirectly('_propagate', data, channel, true);
        }
      });
    }
  }, {
    key: 'getSlowestPeers',
    value: function getSlowestPeers() {
      var peers = (0, _PeerOptimizer.getDroppablePeers)();
    }
  }, {
    key: 'leaveRoom',
    value: function leaveRoom() {
      if (this.roomName) {
        this.connection.emit('leave');
        while (this.webrtc.peers.length) {
          this.webrtc.peers[0].end();
        }
        this.emit('leftRoom', this.roomName);
        this.roomName = undefined;
      }
    }
  }, {
    key: 'disconnect',
    value: function disconnect() {
      this.connection.disconnect();
      delete this.connection;
    }
  }, {
    key: 'handlePeerStreamAdded',
    value: function handlePeerStreamAdded(peer) {
      var self = this;

      this.emit('peerStreamAdded', peer.stream, peer);

      // send our mute status to new peer if we're muted
      // currently called with a small delay because it arrives before
      // the video element is created otherwise (which happens after
      // the async setRemoteDescription-createAnswer)
      setTimeout(function () {
        if (!self.webrtc.isAudioEnabled()) {
          peer.send('mute', { name: 'audio' });
        }
        if (!self.webrtc.isVideoEnabled()) {
          peer.send('mute', { name: 'video' });
        }
      }, 250);
    }
  }, {
    key: 'handlePeerStreamRemoved',
    value: function handlePeerStreamRemoved(peer) {
      // (this.config.media.video) this.emit('peerStreamRemoved', peer);
    }
  }, {
    key: 'getId',
    value: function getId(peer) {
      // eslint-disable-line
      return [peer.id, peer.type, peer.broadcaster ? 'broadcasting' : 'incoming'].join('_');
    }
  }, {
    key: 'getMyId',
    value: function getMyId() {
      return this.id;
    }
  }, {
    key: 'getContainerId',
    value: function getContainerId(peer) {
      return 'container_' + this.getId(peer);
    }

    // set volume on video tag for all peers takse a value between 0 and 1

  }, {
    key: 'setVolumeForAll',
    value: function setVolumeForAll(volume) {
      this.webrtc.peers.forEach(function (peer) {
        if (peer.videoEl) peer.videoEl.volume = volume;
      });
    }
  }, {
    key: 'getClients',
    value: function getClients(callback) {
      this.connection.emit('getClients', this.roomName, function (err, clients) {
        if (callback) callback(err, clients.clients);
      });
    }
  }, {
    key: 'joinRoom',
    value: function joinRoom(name, cb) {
      var _this5 = this;

      var self = this;
      this.roomName = name;
      this.connection.emit('join', name, function (err, roomDescription) {
        if (err) {
          self.emit('error', err);
        } else {
          var id = void 0;
          var client = void 0;
          var type = void 0;
          var peer = void 0;

          _this5.roomCount = Object.keys(roomDescription.clients).length;
          // console.log(roomDescription);
          _this5.id = roomDescription.you;
          (0, _PeerOptimizer.addNode)(_this5.id);
          _this5.unconnectivePeers[_this5.id] = true;
          var _iteratorNormalCompletion = true;
          var _didIteratorError = false;
          var _iteratorError = undefined;

          try {
            for (var _iterator = Object.keys(roomDescription.clients).reverse().filter(function (item) {
              return item !== _this5.id;
            })[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              id = _step.value;

              client = roomDescription.clients[id];
              for (type in client) {
                if (client[type]) {
                  var peerCount = _this5.webrtc.getPeers().length;
                  if (_this5.config.dataOnly && _this5.config.network.maxPeers > 0 && (peerCount >= _this5.config.network.minPeers || peerCount >= _this5.config.network.maxPeers)) {
                    break;
                  }
                  peer = self.webrtc.createPeer({
                    id: id,
                    type: type,
                    enableDataChannels: self.config.enableDataChannels && type !== 'screen',
                    receiveMedia: {
                      offerToReceiveAudio: type !== 'screen' && !_this5.config.dataOnly && _this5.config.receiveMedia.offerToReceiveAudio ? 1 : 0,
                      offerToReceiveVideo: !_this5.config.dataOnly && self.config.receiveMedia.offerToReceiveVideo ? 1 : 0
                    }
                  });
                  _this5.sendPing(peer, peer.id, true);
                }
              }
            }
          } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion && _iterator.return) {
                _iterator.return();
              }
            } finally {
              if (_didIteratorError) {
                throw _iteratorError;
              }
            }
          }
        }

        if (cb) cb(err, roomDescription);
        self.emit('joinedRoom', name);
      });
    }
  }, {
    key: 'startLocalVideo',
    value: function startLocalVideo() {
      var _this6 = this;

      var self = this;
      this.webrtc.start(this.config.media, function (err, stream) {
        if (err) {
          self.emit('localMediaError', err);
        } else {
          (0, _attachmediastream2.default)(stream, _this6.config.localVideoEl, _this6.config.localVideo);
        }
      });
    }
  }, {
    key: 'attachStream',
    value: function attachStream(stream, el, opts) {
      // eslint-disable-line
      var options = {
        autoplay: true,
        muted: false,
        mirror: true,
        audio: false
      };
      (0, _attachmediastream2.default)(stream, el, opts || options);
    }
  }, {
    key: 'setLocalVideo',
    value: function setLocalVideo(element) {
      this.config.localVideoEl = element;
    }
  }, {
    key: 'stopLocalVideo',
    value: function stopLocalVideo() {
      this.webrtc.stop();
    }
  }, {
    key: 'quit',
    value: function quit() {
      this.stopLocalVideo();
      this.leaveRoom();
      this.disconnect();
    }
  }, {
    key: 'testReadiness',
    value: function testReadiness() {
      var self = this;
      if (this.sessionReady) {
        if (this.config.dataOnly || !this.config.media.video && !this.config.media.audio) {
          self.emit('ready', self.connection.getSessionid());
        } else if (this.webrtc.localStreams.length > 0) {
          self.emit('ready', self.connection.getSessionid());
        }
      }
    }
  }, {
    key: 'connectToPeer',
    value: function connectToPeer(peerId, client) {
      var type = void 0;
      var peer = void 0;
      for (type in client) {
        if (client[type]) {
          var peerCount = this.webrtc.getPeers().length;
          if (this.config.network.maxPeers > 0 && peerCount >= this.config.network.maxPeers) {
            break;
          }
          peer = this.webrtc.createPeer({
            id: peerId,
            type: type,
            enableDataChannels: this.config.enableDataChannels && type !== 'screen',
            receiveMedia: {
              offerToReceiveAudio: type !== 'screen' && !this.config.dataOnly && this.config.receiveMedia.offerToReceiveAudio ? 1 : 0,
              offerToReceiveVideo: !this.config.dataOnly && this.config.receiveMedia.offerToReceiveVideo ? 1 : 0
            }
          });
          this.sendPing(peer, peerId, true);
        }
      }
    }
  }, {
    key: 'createRoom',
    value: function createRoom(name, cb) {
      this.roomName = name;
      if (arguments.length === 2) {
        this.connection.emit('create', name, cb);
      } else {
        this.connection.emit('create', name);
      }
    }
  }, {
    key: 'sendFile',
    value: function sendFile() {
      if (!_webrtcsupport2.default.dataChannel) {
        return this.emit('error', new Error('DataChannelNotSupported'));
      }
    }
  }]);

  return LioWebRTC;
}(_wildemitter2.default);

exports.default = LioWebRTC;
