// the-button/shim/esm/TheButtonGroup.js
'use strict';

import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import c from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { eventHandlersFor, htmlAttributesFor } from 'the-component-util';
/**
 * Group of buttons
 */

var TheButtonGroup =
/*#__PURE__*/
function (_React$Component) {
  _inherits(TheButtonGroup, _React$Component);

  function TheButtonGroup() {
    _classCallCheck(this, TheButtonGroup);

    return _possibleConstructorReturn(this, _getPrototypeOf(TheButtonGroup).apply(this, arguments));
  }

  _createClass(TheButtonGroup, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var align = props.align,
          children = props.children,
          className = props.className,
          collapsed = props.collapsed,
          nowrap = props.nowrap;
      return React.createElement("div", _extends({}, htmlAttributesFor(props, {
        except: ['className']
      }), eventHandlersFor(props, {
        except: []
      }), {
        className: c('the-button-group', className, {
          'the-button-group-center': align === 'center',
          'the-button-group-collapsed': collapsed,
          'the-button-group-left': align === 'left',
          'the-button-group-nowrap': nowrap,
          'the-button-group-right': align === 'right'
        })
      }), children);
    }
  }]);

  return TheButtonGroup;
}(React.Component);

TheButtonGroup.propTypes = {
  /** collapsed style */

  /** Direction to align */
  align: PropTypes.oneOf(['left', 'right', 'center']),
  collapsed: PropTypes.bool
};
TheButtonGroup.defaultProps = {
  align: 'center',
  collapsed: false
};
TheButtonGroup.displayName = 'TheButtonGroup';
export default TheButtonGroup;
