// the-button/shim/esm/TheButton.js
'use strict';

import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import c from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { unlessProduction } from 'the-check';
import { textColorFor } from 'the-color';
import { eventHandlersFor, htmlAttributesFor } from 'the-component-util';
import { TheIcon } from 'the-icon';
import { TheLink } from 'the-link';
import TheButtonGroup from './TheButtonGroup';
/**
 * Button of the-components
 */

var TheButton =
/*#__PURE__*/
function (_React$Component) {
  _inherits(TheButton, _React$Component);

  _createClass(TheButton, null, [{
    key: "Next",
    value: function Next(props) {
      return React.createElement(TheButton, _extends({
        iconRight: TheButton.NEXT_ICON
      }, props));
    }
  }, {
    key: "Prev",
    value: function Prev(props) {
      return React.createElement(TheButton, _extends({
        icon: TheButton.PREV_ICON
      }, props));
    }
  }, {
    key: "Spinner",
    value: function Spinner() {
      return React.createElement("span", {
        className: "the-button-spinner"
      }, React.createElement(TheIcon.Spin, {
        className: "the-button-spinner-icon"
      }));
    }
  }]);

  function TheButton(props) {
    var _this;

    _classCallCheck(this, TheButton);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(TheButton).call(this, props));
    _this.handleKeyDown = _this.handleKeyDown.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.handleClick = _this.handleClick.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.lastClickedAt = null;
    return _this;
  }

  _createClass(TheButton, [{
    key: "handleClick",
    value: function handleClick(e) {
      var _this$props = this.props,
          minInterval = _this$props.minInterval,
          onClick = _this$props.onClick,
          onSubmit = _this$props.onSubmit;

      if (this.lastClickedAt) {
        var sinceLast = new Date() - this.lastClickedAt;
        var tooSoon = sinceLast < minInterval;

        if (tooSoon) {
          unlessProduction(function () {
            console.warn("[TheButton] Prevent double click ( minInterval: ".concat(minInterval, "ms, actual: ").concat(sinceLast, "ms )"));
          });
          return;
        }
      }

      onClick && onClick(e);
      onSubmit && onSubmit();
      this.lastClickedAt = new Date();
    }
  }, {
    key: "handleKeyDown",
    value: function handleKeyDown(e) {
      var _this$props2 = this.props,
          onClick = _this$props2.onClick,
          onKeyDown = _this$props2.onKeyDown,
          onSubmit = _this$props2.onSubmit;

      switch (e.keyCode) {
        case 32:
          // Space key
          e.preventDefault();
          onSubmit && onSubmit();
          onClick && onClick();
          break;

        default:
          break;
      }

      onKeyDown && onKeyDown(e);
    }
  }, {
    key: "render",
    value: function render() {
      var props = this.props;
      var children = props.children,
          className = props.className,
          color = props.color,
          danger = props.danger,
          disabled = props.disabled,
          floated = props.floated,
          icon = props.icon,
          iconRight = props.iconRight,
          large = props.large,
          largeIcon = props.largeIcon,
          light = props.light,
          primary = props.primary,
          rounded = props.rounded,
          simple = props.simple,
          small = props.small,
          spinning = props.spinning,
          _props$style = props.style,
          style = _props$style === void 0 ? {} : _props$style,
          target = props.target,
          text = props.text,
          to = props.to,
          wide = props.wide;
      var A = to ? TheLink : 'a';
      var applyColor = color && !danger && !disabled;

      if (applyColor && style) {
        style.borderColor = color;

        if (primary) {
          style.backgroundColor = color;
          style.color = textColorFor(color);
        } else {
          style.color = color;
        }
      }

      return React.createElement(A, _extends({}, htmlAttributesFor(props, {
        except: ['className', 'style', 'icon', 'to', 'color', 'iconRight']
      }), eventHandlersFor(props, {
        except: ['onClick']
      }), {
        "aria-busy": spinning,
        className: c('the-button', className, {
          'the-button-danger': danger,
          'the-button-disabled': disabled,
          'the-button-floated': floated,
          'the-button-large': large,
          'the-button-light': light,
          'the-button-primary': primary,
          'the-button-rounded': rounded,
          'the-button-simple': simple,
          'the-button-small': small,
          'the-button-spinning': spinning,
          'the-button-vertical': !!largeIcon,
          'the-button-wide': wide
        }),
        onClick: this.handleClick
      }, {
        style: style,
        target: target,
        to: to
      }, {
        href: !to && props.href || 'javascript:void(0)',
        onKeyDown: this.handleKeyDown
      }), React.createElement("span", {
        className: "the-button-inner"
      }, spinning && React.createElement(TheButton.Spinner, null), largeIcon && React.createElement(TheIcon, {
        className: c('the-button-large-icon', largeIcon)
      }), icon && React.createElement(TheIcon, {
        className: c('the-button-icon', icon)
      }), text && React.createElement("span", {
        className: c('the-button-text')
      }, text), children, iconRight && React.createElement(TheIcon, {
        className: c('the-button-icon', 'the-button-icon-right', iconRight)
      })));
    }
  }]);

  return TheButton;
}(React.Component);

TheButton.Group = TheButtonGroup;
TheButton.PREV_ICON = 'fas fa-caret-left';
TheButton.NEXT_ICON = 'fas fa-caret-right';
TheButton.propTypes = {
  /** Color theme */
  color: PropTypes.string,

  /** Danger style */
  danger: PropTypes.bool,

  /** Disabled state */
  disabled: PropTypes.bool,

  /** Floated style */
  floated: PropTypes.bool,

  /** Icon class */
  icon: PropTypes.string,

  /** Large style */
  large: PropTypes.bool,

  /** Large icon class */
  largeIcon: PropTypes.string,

  /** Light color */
  light: PropTypes.bool,

  /** Minimum interval */
  minInterval: PropTypes.number,

  /** Handle submit */
  onSubmit: PropTypes.func,

  /** Primary style */
  primary: PropTypes.bool,

  /** Rounded style */
  rounded: PropTypes.bool,

  /** Simple style */
  simple: PropTypes.bool,

  /** Small style */
  small: PropTypes.bool,

  /** Show spinner */
  spinning: PropTypes.bool,

  /** Button text */
  text: PropTypes.string,

  /** Wide style */
  wide: PropTypes.bool
};
TheButton.defaultProps = {
  color: null,
  danger: false,
  disabled: false,
  floated: false,
  icon: null,
  large: false,
  largeIcon: null,
  light: false,
  minInterval: 300,
  onSubmit: null,
  primary: false,
  role: 'button',
  rounded: false,
  simple: false,
  small: false,
  spinning: false,
  text: null,
  wide: false
};
TheButton.displayName = 'TheButton';
export default TheButton;
