// the-spin/shim/TheSpin.js
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _theComponentUtil = require("the-component-util");

var _theIcon = require("the-icon");

var _TheSpinStyle = _interopRequireDefault(require("./TheSpinStyle"));

var SpinnerSizes = ['medium', 'xx-small', 'x-small', 'small', 'large', 'x-large', 'xx-large', 'smaller', 'larger'];
var DEFAULT_THEME = 'C';
var DEFAULT_SIZE = 'large';
/**
 * Spin of the-components
 */

var TheSpin =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(TheSpin, _React$PureComponent);

  function TheSpin() {
    (0, _classCallCheck2.default)(this, TheSpin);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TheSpin).apply(this, arguments));
  }

  (0, _createClass2.default)(TheSpin, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var children = props.children,
          className = props.className,
          cover = props.cover,
          enabled = props.enabled,
          size = props.size,
          theme = props.theme;
      return _react.default.createElement("div", (0, _extends2.default)({}, (0, _theComponentUtil.htmlAttributesFor)(props, {
        except: ['className']
      }), (0, _theComponentUtil.eventHandlersFor)(props, {
        except: []
      }), {
        className: (0, _classnames.default)('the-spin', className, {
          'the-spinner-cover': !!cover,
          'the-spinner-enabled': !!enabled
        })
      }), _react.default.createElement(_theIcon.TheIcon.Spin, {
        className: "the-spin-icon",
        style: {
          fontSize: size
        },
        theme: theme
      }), children);
    }
  }]);
  return TheSpin;
}(_react.default.PureComponent);

TheSpin.Style = _TheSpinStyle.default;
TheSpin.Sizes = SpinnerSizes;
TheSpin.propTypes = {
  /** Shows spin */

  /** show as cover */
  cover: _propTypes.default.bool,
  enabled: _propTypes.default.bool,

  /** Size of spinner icon */
  size: _propTypes.default.oneOf(SpinnerSizes),

  /** Theme of the spinner */
  theme: _propTypes.default.oneOf(Object.keys(_theIcon.TheIcon.SpinIconThemes))
};
TheSpin.defaultProps = {
  cover: false,
  enabled: false,
  size: DEFAULT_SIZE,
  theme: DEFAULT_THEME
};
TheSpin.displayName = 'TheSpin';
var _default = TheSpin;
exports.default = _default;
