// the-spin/shim/withSpin.js
/**
 * HOC for spin
 * @function withSpin
 */
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _asobj = require("asobj");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _TheSpin = _interopRequireDefault(require("./TheSpin"));

/** @lends withSpin */
function withSpin(Component) {
  var WithSpin =
  /*#__PURE__*/
  function (_React$Component) {
    (0, _inherits2.default)(WithSpin, _React$Component);

    function WithSpin() {
      (0, _classCallCheck2.default)(this, WithSpin);
      return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(WithSpin).apply(this, arguments));
    }

    (0, _createClass2.default)(WithSpin, [{
      key: "render",
      value: function render() {
        var props = this.props;
        var children = props.children,
            _props$spinner = props.spinner,
            spinner = _props$spinner === void 0 ? _react.default.createElement(_TheSpin.default, {
          cover: true,
          enabled: spinning,
          size: spinnerSize,
          theme: spinnerTheme
        }) : _props$spinner,
            spinnerSize = props.spinnerSize,
            spinnerTheme = props.spinnerTheme,
            spinning = props.spinning;
        var componentProps = (0, _asobj.clone)(props, {
          without: Object.keys(WithSpin.propTypes)
        });
        return _react.default.createElement(Component, (0, _extends2.default)({}, componentProps, {
          spinner: spinner
        }), children);
      }
    }]);
    return WithSpin;
  }(_react.default.Component);

  WithSpin.propTypes = {
    /** Show loading spinner */

    /** Spinner instance */
    spinner: _propTypes.default.node,

    /** Size of spinner */
    spinnerSize: _propTypes.default.string,

    /** Theme of the spinner */
    spinnerTheme: _propTypes.default.string,
    spinning: _propTypes.default.bool
  };
  return WithSpin;
}

var _default = withSpin;
exports.default = _default;
