// the-check/shim/esm/index.js
// Code generated by coz. DO NOT EDIT.

/**
 * Check utility for the-framework
 * @module the-check
 */
'use strict';

import canTouch_ from './canTouch';
import hasBin_ from './hasBin';
import isBrowser_ from './isBrowser';
import isChrome_ from './isChrome';
import isFirefox_ from './isFirefox';
import isMacOS_ from './isMacOS';
import isProduction_ from './isProduction';
import isiOS_ from './isiOS';
import unlessProduction_ from './unlessProduction'; // `module.exports` overrides these `exports.*`, but still needs them for lebab (https://github.com/lebab/lebab)

export { canTouch_ as canTouch };
export { hasBin_ as hasBin };
export { isBrowser_ as isBrowser };
export { isChrome_ as isChrome };
export { isFirefox_ as isFirefox };
export { isMacOS_ as isMacOS };
export { isProduction_ as isProduction };
export { isiOS_ as isiOS };
export { unlessProduction_ as unlessProduction };
export default {
  canTouch: canTouch_,
  hasBin: hasBin_,
  isBrowser: isBrowser_,
  isChrome: isChrome_,
  isFirefox: isFirefox_,
  isMacOS: isMacOS_,
  isProduction: isProduction_,
  isiOS: isiOS_,
  unlessProduction: unlessProduction_
};
