// the-link/shim/TheLink.js
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactRouterDom = require("react-router-dom");

var _theComponentUtil = require("the-component-util");

var _theIcon = require("the-icon");

var _url = require("url");

var _TheLinkStyle = _interopRequireDefault(require("./TheLinkStyle"));

/**
 * Link of the-components
 */
var TheLink =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(TheLink, _React$Component);

  function TheLink() {
    (0, _classCallCheck2.default)(this, TheLink);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TheLink).apply(this, arguments));
  }

  (0, _createClass2.default)(TheLink, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var activeClassName = props.activeClassName,
          activeStyle = props.activeStyle,
          children = props.children,
          className = props.className,
          color = props.color,
          exact = props.exact,
          icon = props.icon,
          replace = props.replace,
          strict = props.strict,
          _props$style = props.style,
          style = _props$style === void 0 ? {} : _props$style,
          to = props.to;

      if (color && style) {
        style.color = color;
      }

      return _react.default.createElement(TheLink.Link, (0, _extends2.default)({}, (0, _theComponentUtil.htmlAttributesFor)(props, {
        except: ['className', 'style']
      }), (0, _theComponentUtil.eventHandlersFor)(props, {
        except: []
      }), {
        activeClassName: (0, _classnames.default)('the-link-active', activeClassName),
        className: (0, _classnames.default)('the-link', className)
      }, {
        activeStyle: activeStyle,
        exact: exact,
        replace: replace,
        strict: strict,
        style: style,
        to: to
      }), icon && _react.default.createElement(_theIcon.TheIcon, {
        className: (0, _classnames.default)('the-link-icon', icon)
      }), children);
    }
  }], [{
    key: "Link",
    value: function Link(props) {
      var children = props.children,
          className = props.className,
          to = props.to;

      var _ref = to && typeof to === 'string' && (0, _url.parse)(to) || {},
          protocol = _ref.protocol;

      if (protocol) {
        // External link
        return _react.default.createElement("a", (0, _extends2.default)({}, (0, _theComponentUtil.htmlAttributesFor)(props, {
          except: ['className', 'href']
        }), (0, _theComponentUtil.eventHandlersFor)(props, {
          except: []
        }), {
          className: (0, _classnames.default)('the-link-external', className),
          href: to
        }), children);
      }

      return _react.default.createElement(_reactRouterDom.NavLink, props, children);
    }
  }]);
  return TheLink;
}(_react.default.Component);

TheLink.Style = _TheLinkStyle.default;
TheLink.propTypes = {
  /** Class name for active state */
  activeClassName: _propTypes.default.string,

  /** Style for active state */
  activeStyle: _propTypes.default.object,

  /** Color theme */
  color: _propTypes.default.string,

  /** Exact path */
  exact: _propTypes.default.bool,

  /** Icon class name */
  icon: _propTypes.default.string,

  /** String path */
  strict: _propTypes.default.bool,

  /** Link to */
  to: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]).isRequired
};
TheLink.defaultProps = {
  activeClassName: null,
  activeStyle: {},
  color: null,
  exact: false,
  icon: null,
  replace: false,
  role: 'link',
  strict: false,
  to: null
};
TheLink.displayName = 'TheLink';
var _default = TheLink;
exports.default = _default;
