// the-link/shim/TheLinkStyle.js
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _theStyle = require("the-style");

/** Style for TheLink */
var TheLinkStyle = function TheLinkStyle(_ref) {
  var className = _ref.className,
      id = _ref.id,
      options = _ref.options;
  return _react.default.createElement(_theStyle.TheStyle, (0, _extends2.default)({
    id: id
  }, {
    className: (0, _classnames.default)('the-link-style', className),
    styles: TheLinkStyle.data(options)
  }));
};

TheLinkStyle.displayName = 'TheLinkStyle';
TheLinkStyle.propTypes = {
  /** Style options */
  options: _propTypes.default.object
};
TheLinkStyle.defaultProps = {
  options: {}
};

TheLinkStyle.data = function (options) {
  var ThemeValues = _theStyle.TheStyle.ThemeValues;
  var _options$dominantColo = options.dominantColor,
      dominantColor = _options$dominantColo === void 0 ? ThemeValues.dominantColor : _options$dominantColo;
  return {
    '.the-link': {
      color: dominantColor,
      display: 'inline-block',
      padding: '0px 4px'
    },
    '.the-link-active': {
      textDecoration: 'none'
    }
  };
};

var _default = TheLinkStyle;
exports.default = _default;
